File: memory.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// MemoryManagerImpl manages Prime's memory files
type MemoryManagerImpl struct {
	memoryDir     string
	longTermFile  string
	shortTermFile string
}

// NewMemoryManager creates a new memory manager
func NewMemoryManager(memoryDir string) *MemoryManagerImpl {
	return &MemoryManagerImpl{
		memoryDir:     memoryDir,
		longTermFile:  filepath.Join(memoryDir, "long_term.md"),
		shortTermFile: filepath.Join(memoryDir, "short_term.md"),
	}
}

// Initialize initializes memory files if they don't exist
func (m *MemoryManagerImpl) Initialize() error {
	// Create memory directory if it doesn't exist
	if err := os.MkdirAll(m.memoryDir, 0755); err != nil {
		return fmt.Errorf("failed to create memory directory: %v", err)
	}

	// Initialize long-term memory file if it doesn't exist
	if err := m.initializeFile(m.longTermFile, "Long-term"); err != nil {
		return err
	}

	// Initialize short-term memory file if it doesn't exist
	if err := m.initializeFile(m.shortTermFile, "Short-term"); err != nil {
		return err
	}

	return nil
}

func (m *MemoryManagerImpl) initializeFile(path string, memoryType string) error {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		content := fmt.Sprintf("# Prime %s Memory\n\n", memoryType)
		if err := os.WriteFile(path, []byte(content), 0644); err != nil {
			return fmt.Errorf("failed to initialize %s memory file: %v", strings.ToLower(memoryType), err)
		}
	}
	return nil
}

// AddMemory adds a memory entry to the specified memory type
func (m *MemoryManagerImpl) AddMemory(memoryType, category, content string) error {
	var memoryFile string
	switch strings.ToLower(memoryType) {
	case "long", "long_term":
		memoryFile = m.longTermFile
	case "short", "short_term":
		memoryFile = m.shortTermFile
	default:
		return fmt.Errorf("invalid memory type: %s", memoryType)
	}

	// Ensure memory directory and files exist
	if err := m.Initialize(); err != nil {
		return err
	}

	// Read existing memory content
	memoryContent, err := os.ReadFile(memoryFile)
	if err != nil {
		return fmt.Errorf("failed to read memory file: %v", err)
	}

	lines := strings.Split(string(memoryContent), "\n")
	var updatedContent strings.Builder
	categoryFound := false
	categoryProcessed := false

	// Write header and find category location
	for i, line := range lines {
		if strings.HasPrefix(line, fmt.Sprintf("## %s", category)) {
			categoryFound = true
			updatedContent.WriteString(line + "\n")
			timestamp := time.Now().Format("2006-01-02 15:04:05")
			updatedContent.WriteString(fmt.Sprintf("- %s (added: %s)\n", content, timestamp))
			categoryProcessed = true
		} else if categoryProcessed && strings.HasPrefix(line, "## ") {
			// We've found the next category after processing our target
			categoryProcessed = false
			updatedContent.WriteString(line + "\n")
		} else {
			updatedContent.WriteString(line)
			if i < len(lines)-1 {
				updatedContent.WriteString("\n")
			}
		}
	}

	// If category wasn't found, add it at the end
	if !categoryFound {
		if !strings.HasSuffix(updatedContent.String(), "\n") {
			updatedContent.WriteString("\n")
		}
		updatedContent.WriteString(fmt.Sprintf("## %s\n", category))
		timestamp := time.Now().Format("2006-01-02 15:04:05")
		updatedContent.WriteString(fmt.Sprintf("- %s (added: %s)\n", content, timestamp))
	}

	// Write updated content back to file
	if err := os.WriteFile(memoryFile, []byte(updatedContent.String()), 0644); err != nil {
		return fmt.Errorf("failed to write updated memory file: %v", err)
	}

	return nil
}

// ReadMemory reads memory content
func (m *MemoryManagerImpl) ReadMemory(memoryType string) (string, error) {
	// Ensure memory files exist
	if err := m.Initialize(); err != nil {
		return "", err
	}

	switch strings.ToLower(memoryType) {
	case "long", "long_term":
		content, err := os.ReadFile(m.longTermFile)
		if err != nil {
			return "", fmt.Errorf("failed to read long-term memory file: %v", err)
		}
		return string(content), nil

	case "short", "short_term":
		content, err := os.ReadFile(m.shortTermFile)
		if err != nil {
			return "", fmt.Errorf("failed to read short-term memory file: %v", err)
		}
		return string(content), nil

	case "", "all":
		// Combine both memory types
		shortTerm, err := os.ReadFile(m.shortTermFile)
		if err != nil {
			return "", fmt.Errorf("failed to read short-term memory file: %v", err)
		}

		longTerm, err := os.ReadFile(m.longTermFile)
		if err != nil {
			return "", fmt.Errorf("failed to read long-term memory file: %v", err)
		}

		return fmt.Sprintf("%s\n\n%s", string(shortTerm), string(longTerm)), nil

	default:
		return "", fmt.Errorf("invalid memory type: %s", memoryType)
	}
}

// ClearShortTermMemory clears short-term memory
func (m *MemoryManagerImpl) ClearShortTermMemory() error {
	content := "# Prime Short-term Memory\n\n"
	if err := os.WriteFile(m.shortTermFile, []byte(content), 0644); err != nil {
		return fmt.Errorf("failed to clear short-term memory file: %v", err)
	}
	return nil
}

// SearchMemory searches memory for content
func (m *MemoryManagerImpl) SearchMemory(query string, memoryType string) ([]MemoryEntry, error) {
	memoryContent, err := m.ReadMemory(memoryType)
	if err != nil {
		return nil, err
	}

	queryLower := strings.ToLower(query)
	var results []MemoryEntry

	currentCategory := ""
	currentType := memoryType
	if currentType == "" {
		currentType = "all"
	}

	for _, line := range strings.Split(memoryContent, "\n") {
		if strings.HasPrefix(line, "# Prime ") {
			// Memory type header
			if strings.Contains(line, "Long-term") {
				currentType = "long"
			} else if strings.Contains(line, "Short-term") {
				currentType = "short"
			}
		} else if strings.HasPrefix(line, "## ") {
			// Category header
			currentCategory = strings.TrimPrefix(line, "## ")
		} else if strings.HasPrefix(line, "- ") && strings.Contains(strings.ToLower(line), queryLower) {
			// Entry that matches query
			results = append(results, MemoryEntry{
				MemoryType: currentType,
				Category:   currentCategory,
				Content:    strings.TrimPrefix(line, "- "),
			})
		}
	}

	return results, nil
}

// GetCategories gets all categories from memory
func (m *MemoryManagerImpl) GetCategories(memoryType string) ([]string, error) {
	memoryContent, err := m.ReadMemory(memoryType)
	if err != nil {
		return nil, err
	}

	var categories []string
	for _, line := range strings.Split(memoryContent, "\n") {
		if strings.HasPrefix(line, "## ") {
			categories = append(categories, strings.TrimPrefix(line, "## "))
		}
	}

	return categories, nil
}

File: src/memory.rs
use anyhow::{anyhow, Context, Result};
use std::fs;
use std::io::Write;
use std::path::PathBuf;
use chrono::Utc;
 

/// Manages long-term and short-term memory for the assistant
#[derive(Debug, Clone)]
pub struct MemoryManager {
    memory_dir: PathBuf,
}

impl MemoryManager {
    /// Creates a new MemoryManager, ensuring required files exist.
    pub fn new(memory_dir: PathBuf) -> Result<Self> {
        fs::create_dir_all(&memory_dir)
            .with_context(|| format!("Failed to create memory directory at {}", memory_dir.display()))?;

        for mem_file in ["long_term.md", "short_term.md"] {
            let file_path = memory_dir.join(mem_file);
            if !file_path.exists() {
                let header = format!("# Prime {} Memory\n\n(This file is for notes. The AI will read this.)",
                    if mem_file == "long_term.md" { "Long-term" } else { "Short-term" });
                fs::write(&file_path, header)
                    .with_context(|| format!("Failed to create initial memory file at {}", file_path.display()))?;
            }
        }
        Ok(Self { memory_dir })
    }

    /// Reads memory content from the specified file (or both if none specified)
    pub fn read_memory(&self, memory_type: Option<&str>) -> Result<String> {
        let mut memory_content = String::new();
        match memory_type {
            Some("long_term") => {
                let content = self.read_file("long_term.md")?;
                memory_content.push_str("## Long-term Memory\n");
                memory_content.push_str(&content);
            }
            Some("short_term") => {
                let content = self.read_file("short_term.md")?;
                memory_content.push_str("## Short-term Memory\n");
                memory_content.push_str(&content);
            }
            None => {
                let long_term = self.read_file("long_term.md").unwrap_or_default();
                let short_term = self.read_file("short_term.md").unwrap_or_default();
                memory_content.push_str("\n<LONG_TERM_MEMORY>\n");
                memory_content.push_str(long_term.trim());
                memory_content.push_str("\n</LONG_TERM_MEMORY>\n");
                memory_content.push_str("\n<SHORT_TERM_MEMORY>\n");
                memory_content.push_str(short_term.trim());
                memory_content.push_str("\n</SHORT_TERM_MEMORY>\n");
            }
            Some(other) => return Err(anyhow!("Invalid memory type '{}' specified", other)),
        }
        Ok(memory_content)
    }
    
    /// Writes content to the specified memory type
    pub fn write_memory(&self, memory_type: &str, content: &str) -> Result<()> {
        let file_name = match memory_type {
            "long_term" => "long_term.md",
            "short_term" => "short_term.md",
            _ => return Err(anyhow!("Invalid memory type '{}' specified", memory_type)),
        };
        
        let file_path = self.memory_dir.join(file_name);
        let timestamp = Utc::now();
        let entry = format!("\n## Entry ({})\n{}\n", timestamp, content);
        
        fs::OpenOptions::new()
            .append(true)
            .open(&file_path)
            .with_context(|| format!("Failed to open memory file for writing: {}", file_path.display()))
            .and_then(|mut file| {
                file.write_all(entry.as_bytes())
                    .map_err(|e| anyhow::anyhow!("Failed to write to memory file: {}", e))
            })
            .with_context(|| format!("Failed to write to memory file: {}", file_path.display()))
    }
    
    /// Clears the specified memory type
    pub fn clear_memory(&self, memory_type: &str) -> Result<()> {
        let file_name = match memory_type {
            "long_term" => "long_term.md",
            "short_term" => "short_term.md",
            _ => return Err(anyhow!("Invalid memory type '{}' specified", memory_type)),
        };
        
        let file_path = self.memory_dir.join(file_name);
        let header = format!("# Prime {} Memory\n\n(This file is for notes. The AI will read this.)",
            if file_name == "long_term.md" { "Long-term" } else { "Short-term" });
        
        fs::write(&file_path, header)
            .with_context(|| format!("Failed to clear memory file: {}", file_path.display()))
    }
    
    /// Helper to read a specific memory file
    fn read_file(&self, file_name: &str) -> Result<String> {
        let file_path = self.memory_dir.join(file_name);
        fs::read_to_string(&file_path)
            .with_context(|| format!("Failed to read memory file: {}", file_path.display()))
    }
}